internal/engine/selectors/selectors.go

```diff
@@ -0,0 +1,372 @@
+type repoEnvFactory struct{}
+
+func (_ *repoEnvFactory) createEnv() (*cel.Env, error) {
+	return newEnvForEntity(
+		"repository",
+		&minderv1.SelectorEntity_SelectorRepository{},
+		"minder.v1.SelectorEntity.SelectorRepository")
+}
+
+// artifactEnvFactory is a factory for creating a CEL environment
+// for the SelectorArtifact type representing an artifact
+type artifactEnvFactory struct{}
+
+func (_ *artifactEnvFactory) createEnv() (*cel.Env, error) {
+	return newEnvForEntity(
+		"artifact",
+		&minderv1.SelectorEntity_SelectorArtifact{},
+		"minder.v1.SelectorEntity.SelectorArtifact")
+}
+
+// newEnvForEntity creates a new CEL environment for an entity. All environments are allowed to
+// use the generic "entity" variable plus the specific entity type is also declared as variable
+// with the appropriate type.
+func newEnvForEntity(varName string, typ any, typName string) (*cel.Env, error) {
+	entityPtr := &minderv1.SelectorEntity{}
+
+	env, err := cel.NewEnv(
+		cel.Types(typ), cel.Types(&minderv1.SelectorEntity{}),
+		cel.Declarations(
+			decls.NewVar("entity",
+				decls.NewObjectType(string(entityPtr.ProtoReflect().Descriptor().FullName())),
+			),
+			decls.NewVar(varName,
+				decls.NewObjectType(typName),
+			),
+		),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create CEL environment for %s: %v", varName, err)
+	}
+
+	return env, nil
+}
+
+type compiledSelector = cel.Program
+
+// SelectionBuilder is an interface for creating Selections (a collection of compiled CEL expressions)
+// for an entity type. This is what the user of this module uses. The interface makes it easier to pass
+// mocks by the user of this module.
+type SelectionBuilder interface {
+	NewSelectionFromProfile(minderv1.Entity, []*minderv1.Profile_Selector) (Selection, error)
+}
+
+// Env is a struct that holds the CEL environments for each entity type and the factories for creating
+type Env struct {
+	// entityEnvs is a map of entity types to their respective CEL environments. We keep them cached
+	// and lazy-initialize on first use
+	entityEnvs map[minderv1.Entity]*entityEnvCache
+	// factories is a map of entity types to their respective factories for creating CEL environments
+	factories map[minderv1.Entity]celEnvFactory
+}
+
+// entityEnvCache is a struct that holds a CEL environment for lazy-initialization. Since the initialization
+// is done only once, we also keep track of the error
+type entityEnvCache struct {
+	once sync.Once
+	env  *cel.Env
+	err  error
+}
+
+// NewEnv creates a new Env struct with the default factories for each entity type. The factories
+// are used on first access to create the CEL environments for each entity type.
+func NewEnv() (*Env, error) {
+	factoryMap := map[minderv1.Entity]celEnvFactory{
+		minderv1.Entity_ENTITY_UNSPECIFIED:  &genericEnvFactory{},
+		minderv1.Entity_ENTITY_REPOSITORIES: &repoEnvFactory{},
+		minderv1.Entity_ENTITY_ARTIFACTS:    &artifactEnvFactory{},
```

Comment:
**nit:** I tend to prefer function types to interfaces when they have a single method, which is usually the case with factories.

(part 2)

```suggestion
		minderv1.Entity_ENTITY_UNSPECIFIED:  genericEnvFactory,
		minderv1.Entity_ENTITY_REPOSITORIES: repoEnvFactory,
		minderv1.Entity_ENTITY_ARTIFACTS:    artifactEnvFactory,
```